server/src/main/java/org/opensearch/indices/recovery/PeerRecoverySourceService.java

```diff
@@ -324,6 +324,7 @@ void awaitEmpty() {
 
         private final class ShardRecoveryContext {
             final Map<RecoverySourceHandler, RemoteRecoveryTargetHandler> recoveryHandlers = new HashMap<>();
+            private final RecoverySourceHandlerFactory recoverySourceHandlerFactory = new RecoverySourceHandlerFactory();
```

Comment:
nit: This can be a singleton.

Reply:
Java does not allow creating static fields within inner classes. It's supported, I guess, from 16/17+.